packages/financial-templates-lib/src/logger/DiscordTicketTransport.ts

```diff
@@ -0,0 +1,94 @@
+// This transport enables winston logging to trigger opening of tickets through Discord bot. The transport configuration
+// requires bot authentication token and a dictionary of channel IDs. In order to send the ticket command this bot also
+// requires that the passed log object contains both message and mrkdwn strings, as well as discordTicketChannel
+// resolving to any of configured channel ID mapping.
+// This transport should be used with Ticket Tool (https://tickettool.xyz/) configured to trigger ticket commands on
+// the resolved channel ID and whitelisting of bot ID. Also the Discord server should be configured to allow the bot
+// to post messages on the ticket opening channel.
+import { Client, GatewayIntentBits } from "discord.js";
+import * as ss from "superstruct";
+import Transport from "winston-transport";
+
+import { removeAnchorTextFromLinks } from "./Formatters";
+import { isDictionary } from "./Logger";
+
+type TransportOptions = ConstructorParameters<typeof Transport>[0];
+
+const Config = ss.object({
+  botToken: ss.string(),
+  channelIds: ss.optional(ss.record(ss.string(), ss.string())),
+});
+// Config object becomes a type
+// {
+//   botToken: string;
+//   channelIds?: Record<string,string>;
+// }
+export type Config = ss.Infer<typeof Config>;
+
+// this turns an unknown ( like json parsed data) into a config, or throws an error
+export function createConfig(config: unknown): Config {
+  return ss.create(config, Config);
+}
+
+// Interface for log info object.
+interface DiscordTicketInfo {
+  message: string;
+  mrkdwn: string;
+  discordTicketChannel: string;
+}
+
+// Type guard for log info object.
+export const isDiscordTicketInfo = (info: unknown): info is DiscordTicketInfo => {
+  if (!isDictionary(info)) return false;
+  return (
+    typeof info.message === "string" && typeof info.mrkdwn === "string" && typeof info.discordTicketChannel === "string"
+  );
+};
+
+export class DiscordTicketTransport extends Transport {
+  private readonly botToken: string;
+  private readonly channelIds: { [key: string]: string };
+
+  private client: Client;
+
+  constructor(winstonOpts: TransportOptions, { botToken, channelIds = {} }: Config) {
+    super(winstonOpts);
+    this.botToken = botToken;
+    this.channelIds = channelIds;
+
+    this.client = new Client({ intents: [GatewayIntentBits.Guilds] });
+  }
+
+  // Note: info must be any because that's what the base class uses.
+  async log(info: any, callback: () => void): Promise<void> {
+    // We only try sending if we have all expected parameters and a matching channel ID.
+    const canSend = isDiscordTicketInfo(info) && info.discordTicketChannel in this.channelIds;
```

Comment:
Would it be an error if a ticket channel is specified in a log message, but it's not in the list of channel ids we parameterized at startup? When would this happen?

Reply:
Now this would skip the log message if there is no matching channel id provided in the configuration. But I think its worth changing this behavior after we have proper error handling added:
- if isDiscordTicketInfo(info) is false, this means the logging application did not intend to send logs through this transport as it does not have expected info fields. This transport might have been attached only due to inherited environment meant for some other bot, so this would be safe just to skip the logging on this DiscordTicketTransport
- if isDiscordTicketInfo(info) is true, this means the logging application expects to open discord tickets, so it must have proper channel ID configured. If its missing, I think its worth throwing an error (once we have proper handler)